src/pages/iou/steps/MoneyRequestAmountPage.js

```diff
@@ -73,342 +72,317 @@ const defaultProps = {
+                setSelection(selectionValueForState);
             }
             return;
         }
-
-        this.setState((prevState) => {
-            const amount = this.addLeadingZero(`${prevState.amount.substring(0, prevState.selection.start)}${key}${prevState.amount.substring(prevState.selection.end)}`);
-            return this.getNewState(prevState, amount);
-        });
-    }
+        const newAmount = addLeadingZero(`${amount.substring(0, selection.start)}${key}${amount.substring(selection.end)}`);
+        const {amount: amountValueForState, selection: selectionValueForState} = getNewState(amount, selection, newAmount);
+        setAmount(amountValueForState);
+        setSelection(selectionValueForState);
+    };
 
     /**
      * Update long press value, to remove items pressing on <
      *
      * @param {Boolean} value - Changed text from user input
      */
-    updateLongPressHandlerState(value) {
-        this.setState({shouldUpdateSelection: !value});
-        if (!value && !this.textInput.isFocused()) {
-            this.textInput.focus();
+    const updateLongPressHandlerState = useCallback((value) => {
+        setShouldUpdateSelection(!value);
+        if (!value && !textInput.current.isFocused()) {
+            textInput.current.focus();
         }
-    }
+    }, []);
 
     /**
      * Update amount on amount change
      * Validate new amount with decimal number regex up to 6 digits and 2 decimal digit
      *
      * @param {String} text - Changed text from user input
      */
-    updateAmount(text) {
-        this.setState((prevState) => {
-            const amount = this.addLeadingZero(this.replaceAllDigits(text, this.props.fromLocaleDigit));
-            return this.getNewState(prevState, amount);
-        });
-    }
-
-    /**
-     * Replaces each character by calling `convertFn`. If `convertFn` throws an error, then
-     * the original character will be preserved.
-     *
-     * @param {String} text
-     * @param {Function} convertFn - `this.props.fromLocaleDigit` or `this.props.toLocaleDigit`
-     * @returns {String}
-     */
-    replaceAllDigits(text, convertFn) {
-        return _.chain([...text])
-            .map((char) => {
-                try {
-                    return convertFn(char);
-                } catch {
-                    return char;
-                }
-            })
-            .join('')
-            .value();
-    }
-
-    navigateBack() {
-        Navigation.goBack(this.isEditing ? ROUTES.getMoneyRequestConfirmationRoute(this.iouType, this.reportID) : null);
-    }
-
-    navigateToCurrencySelectionPage() {
+    const updateAmount = (text) => {
+        const newAmount = addLeadingZero(replaceAllDigits(text, fromLocaleDigit));
+        const {amount: amountValueForState, selection: selectionValueForState} = getNewState(amount, selection, newAmount);
+        setAmount(amountValueForState);
+        setSelection(selectionValueForState);
+    };
```

Comment:
Instances where we use `prevState` can be tricky to handle. I think the correct rewrite here is the following:

```js
    const updateAmount = (text) => {
        const newAmount = addLeadingZero(replaceAllDigits(text, fromLocaleDigit));
        const newAmountWithoutSpaces = stripSpacesFromAmount(newAmount);
        if (!validateAmount(newAmountWithoutSpaces)) {
            setAmount(prevAmount => prevAmount);
            setSelection(prevSelection => ({...prevSelection}));
            return;
        }
        setAmount(prevAmount => {
            setSelection(prevSelection => {
                return getNewSelection(prevSelection, prevAmount.length, newAmountWithoutSpaces.length);
            });
            return stripCommaFromAmount(newAmountWithoutSpaces);
        });
    };
```

Do you have any idea how we can handle the rest? we have 2 occurrences in `updateAmountNumberPad` but they seem more complex because even the `newAmount` variable is based on `prevState` maybe for those 2 we can just access the state directly?

Reply:
I think it will help if we create a function that updates the amount as follow
```js
    const setNewAmount = (newAmount) => {
        const newAmountWithoutSpaces = stripSpacesFromAmount(newAmount);
        if (!validateAmount(newAmountWithoutSpaces)) {
            setAmount(prevAmount => prevAmount);
            setSelection(prevSelection => ({...prevSelection}));
            return;
        }
        setAmount(prevAmount => {
            setSelection(prevSelection => {
                return getNewSelection(prevSelection, prevAmount.length, newAmountWithoutSpaces.length);
            });
            return stripCommaFromAmount(newAmountWithoutSpaces);
        });
    };
```

then use this function directly instead of the `getNewState` e.g. `updateAmount` would be simply (same thing applies for the 2 occurrences in `updateAmountNumberPad`):
```js
    const updateAmount = (text) => {
        const newAmount = addLeadingZero(replaceAllDigits(text, fromLocaleDigit));
        setNewAmount(newAmount);
    };
```